vpr/src/route/rr_graph_storage.h

```diff
@@ -0,0 +1,578 @@
+ * This structure should **only** contain data used in the inner loop of the *
+ * router.  This data is consider "hot" and the router performance is        *
+ * sensitive to layout and size of this "hot" data.
+ *
+ * Cold data should be stored seperately in t_rr_graph_storage.              *
+ *                                                                           *
+ * xlow, xhigh, ylow, yhigh:  Integer coordinates (see route.c for           *
+ *       coordinate system) of the ends of this routing resource.            *
+ *       xlow = xhigh and ylow = yhigh for pins or for segments of           *
+ *       length 1.  These values are used to decide whether or not this      *
+ *       node should be added to the expansion heap, based on things         *
+ *       like whether it's outside the net bounding box or is moving         *
+ *       further away from the target, etc.                                  *
+ * type:  What is this routing resource?                                     *
+ * cost_index: An integer index into the table of routing resource indexed   *
+ *             data t_rr_index_data (this indirection allows quick dynamic   *
+ *             changes of rr base costs, and some memory storage savings for *
+ *             fields that have only a few distinct values).                 *
+ * capacity:   Capacity of this node (number of routes that can use it).     *
+ *                                                                           *
+ * direction: if the node represents a track, this field                     *
+ *            indicates the direction of the track. Otherwise                *
+ *            the value contained in the field should be                     *
+ *            ignored.                                                       *
+ * side: The side of a grid location where an IPIN or OPIN is located.       *
+ *       This field is valid only for IPINs and OPINs and should be ignored  *
+ *       otherwise.                                                          */
+struct alignas(16) t_rr_node_data {
+    int8_t cost_index_ = -1;
+    int16_t rc_index_ = -1;
+
+    int16_t xlow_ = -1;
+    int16_t ylow_ = -1;
+    int16_t xhigh_ = -1;
+    int16_t yhigh_ = -1;
+
+    t_rr_type type_ = NUM_RR_TYPES;
+    union {
+        e_direction direction; //Valid only for CHANX/CHANY
+        e_side side;           //Valid only for IPINs/OPINs
+    } dir_side_;
+
+    uint16_t capacity_ = 0;
+};
+
+/* t_rr_node_ptc_data is cold data is therefore kept seperate from
+ * t_rr_node_data.
+ *
+ * ptc_num:  Pin, track or class number, depending on rr_node type.          *
+ *           Needed to properly draw.                                        */
+struct t_rr_node_ptc_data {
+    union {
+        int16_t pin_num;
+        int16_t track_num;
+        int16_t class_num;
+    } ptc_;
+};
+
+static_assert(sizeof(t_rr_node_data) == 16, "Check t_rr_node_data size");
```

Comment:
This is a key data structure, so fail an assertion if it gets bigger than expected (16 bytes right now), so developers only expand it after careful consideration.

Reply:
Done